peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+to allow defining read-only :class:`typing.TypedDict` items.
+
+This PEP proposes expanding the scope of ``ReadOnly`` to class and protocol
+attributes, as a single concise way to mark them read-only. Some parity changes
+are also made to :data:`typing.Final`.
+
+Akin to :pep:`705`, it makes no Python grammar changes. Correct usage of
+read-only attributes is intended to be enforced only by static type checkers.
+
+
+Motivation
+==========
+
+Python type system lacks a single concise way to mark an :term:`attribute` read-only.
+This feature is common in other object-oriented languages (such as `C# <https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/readonly>`_),
+and is useful for restricting attribute mutation at a type checker level, as well
+as defining a broad interface for structural subtyping.
+
+.. _classes:
+
+Classes
+-------
+
+There are 3 major ways of achieving read-only attributes, honored by type checkers:
+
+* annotating the attribute with :data:`typing.Final`::
+
+    class Foo:
+        number: Final[int]
+
+        def __init__(self, number: int) -> None:
+            self.number = number
+
+
+    class Bar:
+        def __init__(self, number: int) -> None:
+            self.number: Final = number
+
+  - Supported by :mod:`dataclasses` (and type checkers since `typing#1669 <https://github.com/python/typing/pull/1669>`_).
+  - Overriding ``number`` is not possible - the specification of ``Final``
+    imposes the name cannot be overridden in subclasses.
+
+* read-only proxy via ``@property``::
+
+    class Foo:
+        _number: int
+
+        def __init__(self, number: int) -> None:
+            self._number = number
+
+        @property
+        def number(self) -> int:
+            return self._number
+
+  - Overriding ``number`` is possible. *Type checkers disagree about the specific rules*. [#overriding_property]_
+  - Read-only at runtime. [#runtime]_
+  - Requires extra boilerplate.
+  - Supported by :mod:`dataclasses`, but does not compose well - the synthesized
+    ``__init__`` and ``__repr__`` will use ``_number`` as the parameter/attribute name.
+
+* using a "freezing" mechanism, such as :func:`dataclasses.dataclass` or :class:`typing.NamedTuple`::
+
+    @dataclass(frozen=True)
+    class Foo:
+        number: int
+
+
+    class Bar(NamedTuple):
+        number: int
+
+  - Overriding ``number`` is possible in the ``@dataclass`` case.
+  - Read-only at runtime. [#runtime]_
+  - No per-attribute control - these mechanisms apply to the whole class.
+  - Frozen dataclasses incur some runtime overhead.
+  - ``NamedTuple`` is still a ``tuple``. Most classes do not need to inherit
+    indexing, iteration, or concatenation.
+
+.. _protocols:
+
+Protocols
+---------
+
+Paraphrasing `this post <https://github.com/python/typing/discussions/1525>`_,
+there is no way of defining a :class:`~typing.Protocol`, such that the only
+requirements to satisfy are:
+
+1. ``hasattr(obj, name)``
```

Comment:
```suggestion
1. ``hasattr(obj, "name")``
```
Or use `getattr(obj, name)` on the next line. Right now the two are inconsistent.